### ISSUE TITLE
accuracy stop improving in simple convolution network

### ISSUE THREAD

**Issue by zhangchengcsb** [fchollet/keras#1006]
I've been using keras to build convolution neural networks for binary classification. Using the same input data, I've tried to vary the model structure (i.e. filter size, number of filters, number of hidden layer neurons) for better performance. I noticed that for certain models, the training accuracy remains unchanged at a low value through all 50 training epochs. One of the model structure is as follows:

```python

  n_filters=32
  n_conv_row=n_conv_col=400
  weight_init='he_normal'
  img_rows=img_cols=400
  nn_dense=25
  n_classes=2
  loss_func='categorical_crossentropy'
  model = Sequential()
  model.add(Convolution2D(n_filters, n_conv_row, n_conv_col,
                          border_mode='valid',init=weight_init,
                          input_shape=(1, img_rows, img_cols)))
  model.add(Activation('relu'))
  model.add(Flatten())
  model.add(Dense(nn_dense, init=weight_init))
  model.add(Activation('relu'))
  model.add(Dense(n_classes, init=weight_init))
  model.add(Activation('softmax'))
  model.compile(loss=loss_func, optimizer='sgd')

```
I used the whole training set as validation so I can get the training accuracy at the end of each epoch. The output after each epoch looks like:
```
259s - loss: 10.4525 - acc: 0.4903 - val_loss: 11.4543 - val_acc: 0.4473
269s - loss: 11.4543 - acc: 0.4473 - val_loss: 11.4543 - val_acc: 0.4473
251s - loss: 11.4543 - acc: 0.4473 - val_loss: 11.4543 - val_acc: 0.4473
252s - loss: 11.4543 - acc: 0.4473 - val_loss: 11.4543 - val_acc: 0.4473
...
```

I plotted the filter weights of the convolution layer at each epoch and they stayed the same for most of the epochs and had minor changes rarely.
at the beginning of epoch 1:
![epoch1_batch1_layer0_weight_img](https://cloud.githubusercontent.com/assets/7233438/11153096/32e26b80-89fc-11e5-8baa-71a2c2b2ef53.png)
at the beginning of epoch 3:
![epoch3_batch1_layer0_weight_img](https://cloud.githubusercontent.com/assets/7233438/11153097/32ee8bf4-89fc-11e5-8252-6bedd64d4888.png)
at the beginning of epoch 5:
![epoch5_batch1_layer0_weight_img](https://cloud.githubusercontent.com/assets/7233438/11153095/32e1ff56-89fc-11e5-9f39-e5e471a69538.png)

I also tried rmsprop as optimizer, and the accuracy at each epoch was exactly the same 0.4473 . However the filter weights formed different pattern at the end of epoch 1 but remained unchanged after that.

What may I have missed? I tried this model structure directly with Theano and the problem persisted, so I guess this problem was not from keras.

To identify the cause of this problem, I think it would be helpful to track the activation and gradient values of each mini-batch. How do I do these with Keras or Theano? Or what would be the right direction to look into?

Thanks a lot.
Cheng

**Comment by EderSantana** [comment]
are you using default learning rates?

**Comment by zhangchengcsb** [comment]
@EderSantana In the above code I used the default learning rate for SGD which is 0.01. I've also suspected this might cause the problem, so I tried 0.1, 0.3, 0.9 and got slight different patterns in the filters but the same unchanged validation accuracy. As I mentioned I also tried rmsprop with default parameters, only to get the same accuracy.

**Comment by EderSantana** [comment]
I think your learning rates are too large, actually... and try some dropout. Maybe you are getting stuck in high eigenvalue spread regions (narrow valleys where you move a lot without changing much in the overall cost function). How large and correlated is your input data?

**Comment by EderSantana** [comment]
wait, are you using a filters of size 400?

**Comment by zhangchengcsb** [comment]
@EderSantana Thanks a lot for your advice. I tried using learning rate 0.0001 and 0.0003 and the validation accuracy improved over 30 epochs. Now I have validation accuracy (actually training accuracy on the whole data set) at around 0.6 . It seems that the problem was really caused by a large learning rate.

  It's possible that the model got into a valley within a few mini-batches due to large learning rate. Is there a way that I can confirm and/or avoid this? 

  I've not checked my input data for correlation yet. Can I do this by unrolling the input images into vectors and test for correlation between these vectors? What should I do about correlated inputs? Maybe cluster them by correlation and pick one from each cluster?

  The filter size is equal to my input image size, i.e. 400 by 400. My input 'images' are actually not images but grid of values with each position indicating certain feature. So I hope to fix the filter position without striding, and let them learn patterns from my input. I understand this may not be a legitimate way to use convolution networks. I'm trying to run with filter size at 50 by 50, but still trying to figure out a way for them to work with my limited computation resources.

**Comment by EderSantana** [comment]
using a filter of the size of the image is what we mean by fully connected layer. It is much faster. Also, if you want convolutions, try 3x3 and stuff like that... check lenet5 architecture

If you have images and the neighboring pixels are correlated like in regular images you will probably have correlations issues, that was why batch normalization was invented. Give Keras's BatchNormalization layer a try. It seems to be working correctly now.

**Comment by zhangchengcsb** [comment]
@EderSantana Thanks for the comments. I understand the fully connected layer part, a filter with fixed position is indeed a hidden layer neuron. I'll try some different filter sizes.

I tried adding batch normalization layer after each layer in my model, only after the convolution layer, and only after the fully connected layer. It seems only the model with batch normalization after each layer allows gradually improving of accuracy over 50 epochs. However the speed of improvement seems slower than a model with correct learning rate (e.g. 0.0003) and no batch normalization. Is it supposed to perform like this? Thanks.

**Comment by EderSantana** [comment]
nah, its supposed to learn faster with batch normalization, maybe with it you can try a learning rates that are a little bit larger

**Comment by zhangchengcsb** [comment]
@EderSantana Yes, I'm able to use a learning rate of 0.001 with batch normalization for faster training. Thanks for all the help.

**Comment by EderSantana** [comment]
awesome! good to hear that

don't forget to close the issue when you are done.

**Comment by vsuriya93** [comment]
I trying to run a CNN for image classification in CIFAR, but accuracy stops improving after 10% This is my model. 

model=Sequential()

model.add(Convolution2D(32,3,3,border_mode='same',input_shape=(3,32,32)))
model.add(Activation('relu'))
model.add(Convolution2D(32,3,3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(.25))
model.add(Flatten())

model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dropout(.5))

model.add(Dense(10))
model.add(Activation('softmax'))

sgd = SGD(lr=0.003, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy',optimizer=sgd)
model.fit(train,y_train,nb_epoch=15,show_accuracy=True,verbose=1,validation_data=(test,y_test))

I have tried varying the learning rate also. Nothing seems to work.

Train - > (40000,3,32,32)
Test - > (10000,3,32,32)

y_train -> (40000,10) # the classes have been binarized
y_test -> (10000,10) # the classes have been binarized

**Comment by EderSantana** [comment]
did you try adam instead of sgd with decay? That model should not be too bad, I think...

**Comment by fatemaaa** [comment]
I'm trying to use the vgg19 architectures for my model but the validation accuracy I got is constant value around 56% and I used too many values of the learning rate such as 0.1, 0.01, 0.001,0.0001 and all of them gave the same result

**Comment by vinayakumarr** [comment]
How did you plot the filter weights of convolutional layer.

**Comment by vinayakumarr** [comment]
I have a trained model file with three convolutional layer and another model file with 2 convolution and 1 lstm layer. I want to plot the filter weights of each layer. How to do this.

**Comment by ravishk1** [comment]
I am not that much familiar with python, so i am using MATLAB for CNN training. I have a data set of 27,000 images and angles corresponding to that images.
My sample code is : 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
layers = [imageInputLayer([32 32 1])
          convolution2dLayer(5,20)
          reluLayer()
          maxPooling2dLayer(2,'Stride',2)
          fullyConnectedLayer(size(categories(trainAngle)))
          softmaxLayer
          classificationLayer];
      
      
 options = trainingOptions('sgdm', 'MaxEpochs', 20, ...
          'InitialLearnRate', 0.0003);
      
 convnet = trainNetwork(trainZ, trainAngle, layers,options);
 
 resultant_Train = classify(convnet,trainZ); %Training data
 resultant_Valid = classify(convnet,validZ); %Validation data

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

My training accuracy is 70% 
but test accuracy is only 2%;
I am completely blank what to do next. Do you have any suggestion? How can I improve my test accuracy?

**Comment by nhulkund** [comment]
This is a perfect example of overfitting - try using dropout layers, data augmentation, adding noise, or make the model simpler (it is memorizing your own training data).

**Comment by Paulraita** [comment]
I had a same issue. When I was training the CNN, the val_acc never updated. And there was minor updates in acc. but it seemed to vibrate. At last, I found the solution.
The issue was in my training dataset. The size was about 4000. And the size of validation dataset was 500
I thought that this issue came from the network weight initialization in flatten layer.
So I changed the dataset and tested. The result was tremendous.

**This was my first training part.**

```
model.fit_generator(
        train_data,
        steps_per_epoch=4091 // 32,
        epochs=100,
        validation_data=test_data,
        validation_steps=474 // 32,
        verbose=1,
        workers=4)
```

**I changed it as follows:**

```
model.fit_generator(
        test_data,
        steps_per_epoch=474 // 32,
        epochs=25,
        verbose=1,
        workers=4)
```

And for the next training, I saved the model weights and reused for next training.
`model.save_weights('weight_model.h5')`

And initialized network weights with this:
```
# Load model weights
model.load_weights('weight_model.h5')

# Compiling the ANN
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
```

This worked perfectly. I hope this will be helpful. Thanks.

**Comment by cmosquer** [comment]
@Paulraita  Hey! I'm having the same problem that you described. But I don't fully understand your solution. Did you run both codes one after the other? or directly trained with your test_data?
I mean, is this what you run?: 
----------------------------------------------------------------------------
model.fit_generator(
        train_data,
        steps_per_epoch=4091 // 32,
        epochs=100,
        validation_data=test_data,
        validation_steps=474 // 32,
        verbose=1,
        workers=4)
model.fit_generator(
        test_data,
        steps_per_epoch=474 // 32,
        epochs=25,
        verbose=1,
        workers=4)


----------------------------------------
Or did you only fit the test_data at first? I mean, just run: 
model.fit_generator(
        test_data,
        steps_per_epoch=474 // 32,
        epochs=25,
        verbose=1,
        workers=4)
-------------------------------------------------------------------
Thank you very much!

**Comment by Paulraita** [comment]
Hello @cmosquer 
first, I started with small dataset. and saved the model.
I increased the dataset size gradually. In this way, I could overcome this issue.
I hope this will be helpful for you.

**Comment by Paulraita** [comment]
@gauravgola96 
I checked its validation accuracy with separate dataset.
For some solutions, it could be helpful to start from small size of dataset.
And we can't solve all the kind of issues like this with the method I used before.
There are many solutions to overcome this. For example, in my case, batch normalization was one of the solutions. And for your last question, I never changed it. Nowadays I often use callback to monitor the training process. Did you try to change the learning rate?

**Comment by gauravgola96** [comment]
@Paulraita 
Yes, I have changed the learning rate and also used ReduceLROnPlateau callback as well.
When trained my inceptionv3 with first 90 layers freezed and trained it on small dataset of 350 images from 12 classes.I used the weights (trained on small dataset) for training the large dataset afterwards on same network and with same number of freezed layers. The training accuracy increased very much but the validation accuracy still able to cross 60%.

Train data - 7200 images from 12 classes (balanced)
Test data - 350 images from 12 classes
